python/cuml/experimental/explainer/base.py

```diff
@@ -0,0 +1,186 @@
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# Copyright (c) 2020, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import cudf
+import cupy as cp
+import numpy as np
+import pandas
+
+import cuml.common.logger as logger
+from cuml.experimental.explainer.common import get_dtype_from_model_func
+from cuml.experimental.explainer.common import get_handle_from_cuml_model_func
+from cuml.experimental.explainer.common import get_link_fn_from_str_or_fn
+from cuml.experimental.explainer.common import get_tag_from_model_func
+from cuml.experimental.explainer.common import model_func_call
+from cuml.common.input_utils import input_to_cupy_array
+
+
+class SHAPBase():
+    """
+    Base class for SHAP based explainers.
+
+    Parameters
+    ----------
+    model : function
+        Function that takes a matrix of samples (n_samples, n_features) and
+        computes the output for those samples with shape (n_samples). Function
+        must use either CuPy or NumPy arrays as input/output.
+    data : Dense matrix containing floats or doubles.
+        Background dataset. Dense arrays are supported.
+    order : 'F', 'C' or None (default = None)
+        Set to override detection of row ('C') or column ('F') major order,
+        if None it will be attempted to be inferred from model.
+    order_default : 'F' or 'C' (default = 'C')
+        Used when `order` is None. If the order cannot be inferred from the
+        model, then order is set to `order_default`.
+    link : function or str (default = 'identity')
+        The link function used to map between the output units of the
+        model and the SHAP value units.
+    random_state: int, RandomState instance or None (default = None)
+        Seed for the random number generator for dataset creation.
+    gpu_model : bool or None (default = None)
```

Comment:
I would prefer the name `is_gpu_model` since `gpu_model` sounds like it should point to an actual model object

Reply:
good point, it is confusing as it is indeed, will change it